tests/lms/util/test_h.py

```diff
@@ -0,0 +1,142 @@
+            ),
+            # Capital letters, dots and underscores in the full name are preserved.
+            (
+                "Jane_Q._Public",
+                "0ae836b9-7fc9-4060-006f-27b2066ac545",
+                "Jane_Q._Public0ae836b9",
+            ),
+            # Capital letters, dots and underscores in the user_id are preserved too.
+            (
+                "janeqpublic",
+                "0.A_e836b9-7fc9-4060-006f-27b2066ac545",
+                "janeqpublic0.A_e836",
+            ),
+            # Spaces are removed from the full name.
+            (
+                "jane q public",
+                "0ae836b9-7fc9-4060-006f-27b2066ac545",
+                "janeqpublic0ae836b9",
+            ),
+            # Multiple spaces in a row get removed too.
+            (
+                "jane  q   public",
+                "0ae836b9-7fc9-4060-006f-27b2066ac545",
+                "janeqpublic0ae836b9",
+            ),
+            # Leading and trailing spaces get removed too.
+            (
+                " janeqpublic  ",
+                "0ae836b9-7fc9-4060-006f-27b2066ac545",
+                "janeqpublic0ae836b9",
+            ),
+            # Other special characters are also removed from the full name.
+            (
+                "jane-@-public!",
+                "0ae836b9-7fc9-4060-006f-27b2066ac545",
+                "janepublic0ae836b9",
+            ),
+            # Special chars are removed from the userid too.
+            (
+                "janeqpublic",
+                " 0a-@b! 0ae836b9-7fc9-4060-006f-27b2066ac545",
+                "janeqpublic0ab0ae83",
+            ),
+        ],
+    )
+    def test_it_removes_disallowed_chars_from_full_name(
+        self, full_name, user_id, expected_username
+    ):
+        request_params = {"lis_person_name_full": full_name, "user_id": user_id}
+
+        username = generate_username(request_params)
+
+        assert username == expected_username
+
+    def test_it_uses_anonymous_if_the_full_name_is_missing(self):
+        request_params = {"user_id": "0ae836b9-7fc9-4060-006f-27b2066ac545"}
+
+        username = generate_username(request_params)
+
+        assert username == "Anonymous0ae836b9"
+
+    @pytest.mark.parametrize(
+        "full_name", [None, "", "   ", "!$@-"]  # All disallowed characters.
+    )
+    def test_it_uses_anonymous_if_the_full_name_is_empty(self, full_name):
+        request_params = {
+            "lis_person_name_full": full_name,
+            "user_id": "0ae836b9-7fc9-4060-006f-27b2066ac545",
+        }
+
+        username = generate_username(request_params)
+
+        assert username == "Anonymous0ae836b9"
+
+    def test_it_uses_random_characters_if_the_userid_is_missing(self, random):
+        request_params = {"lis_person_name_full": "janeqpublic"}
+
+        username = generate_username(request_params)
+
+        assert username == "janeqpublicrandom_c"
+
+        # Also check that it called the random library properly.
```

Comment:
This feels a little tacked-on. Maybe make it its own actual test?

Reply:
Ok, I've moved this into a separate test